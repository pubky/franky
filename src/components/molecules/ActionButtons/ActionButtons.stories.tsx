import type { Meta, StoryObj } from '@storybook/nextjs-vite';

import { ActionButtons } from './ActionButtons';

const meta: Meta<typeof ActionButtons> = {
  title: 'Molecules/Buttons/ActionButtons',
  component: ActionButtons,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: 'Authentication call-to-action buttons featuring sign in and account creation.',
      },
    },
  },
  args: {
    signInText: 'Sign in',
    createAccountText: 'Create account',
  },
  argTypes: {
    onSignIn: { action: 'sign in clicked' },
    onCreateAccount: { action: 'create account clicked' },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Playground: Story = {
  render: (args) => <ActionButtons {...args} />,
};

export const CustomLabels: Story = {
  args: {
    signInText: 'Log in',
    createAccountText: 'Register',
  },
};

export const CompactLayout: Story = {
  args: {
    className: 'w-full flex-col sm:flex-row gap-2',
  },
  parameters: {
    docs: {
      description: {
        story: 'Demonstrates applying a custom class to adjust layout responsiveness.',
      },
    },
  },
};
