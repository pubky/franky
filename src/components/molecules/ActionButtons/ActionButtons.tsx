import * as Atoms from '@/atoms';
import * as Libs from '@/libs';

interface ActionButtonsProps {
  className?: React.HTMLAttributes<HTMLDivElement>['className'];
  onSignIn?: () => void;
  onCreateAccount?: () => void;
  signInText?: string;
  createAccountText?: string;
}

export function ActionButtons({
  className,
  onSignIn,
  onCreateAccount,
  signInText = 'Sign in',
  createAccountText = 'Create account',
  ...props
}: ActionButtonsProps) {
  return (
    <Atoms.Container className={Libs.cn('gap-3 flex-row sm:items-center', className)} {...props}>
      <Atoms.Button variant="secondary" className="w-[158px] sm:w-auto" size="lg" onClick={onSignIn}>
        <Libs.LogIn className="mr-2 h-4 w-4" />
        {signInText}
      </Atoms.Button>
      <Atoms.Button className="w-[158px] sm:w-auto" size="lg" onClick={onCreateAccount}>
        <Libs.UserRoundPlus className="mr-2 h-4 w-4" />
        {createAccountText}
      </Atoms.Button>
    </Atoms.Container>
  );
}
