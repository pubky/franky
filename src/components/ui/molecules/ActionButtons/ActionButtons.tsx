'use client';

import { LogIn, UserRoundPlus } from 'lucide-react';
import { Button, Container } from '@/components/ui';
import { cn } from '@/libs';

interface ActionButtonsProps {
  className?: React.HTMLAttributes<HTMLDivElement>['className'];
  onSignIn?: () => void;
  onCreateAccount?: () => void;
  signInText?: string;
  createAccountText?: string;
}

export function ActionButtons({
  className,
  onSignIn,
  onCreateAccount,
  signInText = 'Sign in',
  createAccountText = 'Create account',
  ...props
}: ActionButtonsProps) {
  return (
    <Container className={cn('flex-col gap-3 sm:flex-row sm:items-center pt-6', className)} {...props}>
      <Button variant="secondary" className="w-full sm:w-auto order-2 sm:order-1" size="lg" onClick={onSignIn}>
        <LogIn className="mr-2 h-4 w-4" />
        {signInText}
      </Button>
      <Button className="w-full sm:w-auto order-1 sm:order-2" size="lg" onClick={onCreateAccount}>
        <UserRoundPlus className="mr-2 h-4 w-4" />
        {createAccountText}
      </Button>
    </Container>
  );
}
