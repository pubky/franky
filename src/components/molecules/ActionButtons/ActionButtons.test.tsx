import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { ActionButtons } from './ActionButtons';

// Mock @/libs to intercept Libs.LogIn and Libs.UserRoundPlus
vi.mock('@/libs', () => ({
  LogIn: ({ className }: { className?: string }) => (
    <div data-testid="login-icon" className={className}>
      LogIn
    </div>
  ),
  UserRoundPlus: ({ className }: { className?: string }) => (
    <div data-testid="user-plus-icon" className={className}>
      UserRoundPlus
    </div>
  ),
  cn: (...inputs: (string | undefined | null | false)[]) => inputs.filter(Boolean).join(' '),
}));

describe('ActionButtons', () => {
  it('renders both buttons with default text', () => {
    render(<ActionButtons />);

    const signInButton = screen.getByRole('button', { name: /sign in/i });
    const createAccountButton = screen.getByRole('button', { name: /create account/i });

    expect(signInButton).toBeInTheDocument();
    expect(createAccountButton).toBeInTheDocument();

    // Check icons are present
    expect(screen.getByTestId('login-icon')).toBeInTheDocument();
    expect(screen.getByTestId('user-plus-icon')).toBeInTheDocument();
  });

  it('renders with custom button text', () => {
    render(<ActionButtons signInText="Log In" createAccountText="Register" />);

    expect(screen.getByRole('button', { name: /log in/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /register/i })).toBeInTheDocument();
  });

  it('calls onSignIn when sign in button is clicked', () => {
    const mockOnSignIn = vi.fn();
    render(<ActionButtons onSignIn={mockOnSignIn} />);

    const signInButton = screen.getByRole('button', { name: /sign in/i });
    fireEvent.click(signInButton);

    expect(mockOnSignIn).toHaveBeenCalledTimes(1);
  });

  it('calls onCreateAccount when create account button is clicked', () => {
    const mockOnCreateAccount = vi.fn();
    render(<ActionButtons onCreateAccount={mockOnCreateAccount} />);

    const createAccountButton = screen.getByRole('button', { name: /create account/i });
    fireEvent.click(createAccountButton);

    expect(mockOnCreateAccount).toHaveBeenCalledTimes(1);
  });

  it('applies default className structure', () => {
    render(<ActionButtons />);

    const container = screen.getByRole('button', { name: /sign in/i }).parentElement;
    expect(container).toHaveClass('mx-auto', 'w-full', 'flex', 'gap-3', 'flex-row', 'sm:items-center');
  });

  it('applies custom className', () => {
    render(<ActionButtons className="custom-action-buttons" />);

    const container = screen.getByRole('button', { name: /sign in/i }).parentElement;
    expect(container).toHaveClass('custom-action-buttons');
  });

  it('has proper button variants and sizes', () => {
    render(<ActionButtons />);

    const signInButton = screen.getByRole('button', { name: /sign in/i });
    const createAccountButton = screen.getByRole('button', { name: /create account/i });

    // Sign in button should be secondary variant
    expect(signInButton).toHaveAttribute('data-variant', 'secondary');

    // Create account button should be primary (default) variant - no explicit variant means default
    expect(createAccountButton).toHaveAttribute('data-testid', 'button-default');

    // Both buttons should have proper size classes
    expect(signInButton).toHaveClass('h-10');
    expect(createAccountButton).toHaveClass('h-10');
  });

  it('renders icons with proper classes', () => {
    render(<ActionButtons />);

    const loginIcon = screen.getByTestId('login-icon');
    const userPlusIcon = screen.getByTestId('user-plus-icon');

    expect(loginIcon).toHaveClass('mr-2', 'h-4', 'w-4');
    expect(userPlusIcon).toHaveClass('mr-2', 'h-4', 'w-4');
  });

  it('handles both callbacks simultaneously', () => {
    const mockOnSignIn = vi.fn();
    const mockOnCreateAccount = vi.fn();

    render(<ActionButtons onSignIn={mockOnSignIn} onCreateAccount={mockOnCreateAccount} />);

    const signInButton = screen.getByRole('button', { name: /sign in/i });
    const createAccountButton = screen.getByRole('button', { name: /create account/i });

    fireEvent.click(signInButton);
    fireEvent.click(createAccountButton);

    expect(mockOnSignIn).toHaveBeenCalledTimes(1);
    expect(mockOnCreateAccount).toHaveBeenCalledTimes(1);
  });
});
