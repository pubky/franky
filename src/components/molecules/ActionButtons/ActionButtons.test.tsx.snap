// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionButtons - Snapshots > matches snapshot with both callbacks 1`] = `
<div
  class="mx-auto w-full flex-col flex gap-3 flex-row sm:items-center"
  data-testid="container"
>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
    data-variant="secondary"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="login-icon"
    >
      LogIn
    </div>
    Sign in
  </button>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-brand/20 text-brand hover:!bg-brand/30 border-brand h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="user-plus-icon"
    >
      UserRoundPlus
    </div>
    Create account
  </button>
</div>
`;

exports[`ActionButtons - Snapshots > matches snapshot with create account callback only 1`] = `
<div
  class="mx-auto w-full flex-col flex gap-3 flex-row sm:items-center"
  data-testid="container"
>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
    data-variant="secondary"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="login-icon"
    >
      LogIn
    </div>
    Sign in
  </button>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-brand/20 text-brand hover:!bg-brand/30 border-brand h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="user-plus-icon"
    >
      UserRoundPlus
    </div>
    Create account
  </button>
</div>
`;

exports[`ActionButtons - Snapshots > matches snapshot with custom className 1`] = `
<div
  class="mx-auto w-full flex-col flex gap-3 flex-row sm:items-center custom-action-buttons"
  data-testid="container"
>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
    data-variant="secondary"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="login-icon"
    >
      LogIn
    </div>
    Sign in
  </button>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-brand/20 text-brand hover:!bg-brand/30 border-brand h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="user-plus-icon"
    >
      UserRoundPlus
    </div>
    Create account
  </button>
</div>
`;

exports[`ActionButtons - Snapshots > matches snapshot with custom text 1`] = `
<div
  class="mx-auto w-full flex-col flex gap-3 flex-row sm:items-center"
  data-testid="container"
>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
    data-variant="secondary"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="login-icon"
    >
      LogIn
    </div>
    Log In
  </button>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-brand/20 text-brand hover:!bg-brand/30 border-brand h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="user-plus-icon"
    >
      UserRoundPlus
    </div>
    Register
  </button>
</div>
`;

exports[`ActionButtons - Snapshots > matches snapshot with default props 1`] = `
<div
  class="mx-auto w-full flex-col flex gap-3 flex-row sm:items-center"
  data-testid="container"
>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
    data-variant="secondary"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="login-icon"
    >
      LogIn
    </div>
    Sign in
  </button>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-brand/20 text-brand hover:!bg-brand/30 border-brand h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="user-plus-icon"
    >
      UserRoundPlus
    </div>
    Create account
  </button>
</div>
`;

exports[`ActionButtons - Snapshots > matches snapshot with sign in callback only 1`] = `
<div
  class="mx-auto w-full flex-col flex gap-3 flex-row sm:items-center"
  data-testid="container"
>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-secondary text-secondary-foreground hover:bg-secondary/80 h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
    data-variant="secondary"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="login-icon"
    >
      LogIn
    </div>
    Sign in
  </button>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*=\\"size-\\"])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive font-semibold cursor-pointer rounded-full border shadow-xs bg-brand/20 text-brand hover:!bg-brand/30 border-brand h-10 gap-1 px-8 py-7 has-[>svg]:px-6 md:has-[>svg]:px-8 w-[158px] sm:w-auto"
    data-size="lg"
    data-slot="button"
  >
    <div
      class="mr-2 h-4 w-4"
      data-testid="user-plus-icon"
    >
      UserRoundPlus
    </div>
    Create account
  </button>
</div>
`;
